src/test/java/com/google/cloud/spanner/pgadapter/JdbcSimpleModeMockServerTest.java

```diff
@@ -259,6 +508,366 @@ public void testDdl() throws SQLException {
+    List<ExecuteBatchDmlRequest> batchDmlRequests =
+        mockSpanner.getRequestsOfType(ExecuteBatchDmlRequest.class);
+    assertEquals(1, batchDmlRequests.size());
+    ExecuteBatchDmlRequest batchDmlRequest = batchDmlRequests.get(0);
+    assertTrue(batchDmlRequest.getTransaction().hasBegin());
+    assertEquals(INSERT_STATEMENT.getSql(), batchDmlRequest.getStatements(0).getSql());
+    assertEquals(UPDATE_STATEMENT.getSql(), batchDmlRequest.getStatements(1).getSql());
+
+    List<CommitRequest> commitRequests = mockSpanner.getRequestsOfType(CommitRequest.class);
+    assertEquals(1, commitRequests.size());
+  }
+
+  @Test
+  public void testBeginInExplicitTransaction() throws SQLException {
+    String sql =
+        "BEGIN TRANSACTION; INSERT INTO FOO VALUES (1); UPDATE FOO SET BAR=1 WHERE BAZ=2; COMMIT;";
+    try (Connection connection = DriverManager.getConnection(createUrl())) {
+      try (java.sql.Statement statement = connection.createStatement()) {
+        // Start an explicit transaction before executing batch
+        assertFalse(statement.execute("BEGIN; SELECT 1"));
+        assertEquals(0, statement.getUpdateCount());
+
+        assertTrue(statement.getMoreResults());
+        try (ResultSet resultSet = statement.getResultSet()) {
+          assertTrue(resultSet.next());
+          assertEquals(1L, resultSet.getLong(1));
+          assertFalse(resultSet.next());
+        }
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+
+        // The BEGIN statement should be no-op and not cause exception.
+        assertFalse(statement.execute(sql));
+        // The BEGIN statement should not return anything.
+        assertEquals(0, statement.getUpdateCount());
+        // INSERT
+        assertFalse(statement.getMoreResults());
+        assertEquals(1, statement.getUpdateCount());
+        // UPDATE
+        assertFalse(statement.getMoreResults());
+        assertEquals(2, statement.getUpdateCount());
+        // COMMIT
+        assertFalse(statement.getMoreResults());
+        assertEquals(0, statement.getUpdateCount());
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+      }
+    }
+
+    List<ExecuteBatchDmlRequest> batchDmlRequests =
+        mockSpanner.getRequestsOfType(ExecuteBatchDmlRequest.class);
+    assertEquals(1, batchDmlRequests.size());
+    ExecuteBatchDmlRequest batchDmlRequest = batchDmlRequests.get(0);
+    // Verify that the BEGIN statement is no-op
+    assertFalse(batchDmlRequest.getTransaction().hasBegin());
+    assertTrue(batchDmlRequest.getTransaction().hasId());
+    assertEquals(INSERT_STATEMENT.getSql(), batchDmlRequest.getStatements(0).getSql());
+    assertEquals(UPDATE_STATEMENT.getSql(), batchDmlRequest.getStatements(1).getSql());
+
+    List<CommitRequest> commitRequests = mockSpanner.getRequestsOfType(CommitRequest.class);
+    assertEquals(1, commitRequests.size());
+  }
+
```

Comment:
(I've reviewed everything up until here. I'll try to continue later today with the rest of these tests)